metrics-exporter-prometheus/src/exporter/http_listener.rs

```diff
@@ -0,0 +1,131 @@
+use std::{
+    net::{IpAddr, SocketAddr},
+    sync::Arc,
+};
+
+use http_body_util::Full;
+use hyper::{
+    body::{self, Bytes, Incoming},
+    server::conn::http1::Builder as HyperHttpBuilder,
+    service::service_fn,
+    Request, Response, StatusCode,
+};
+use hyper_util::rt::TokioIo;
+use ipnet::IpNet;
+use tokio::net::{TcpListener, TcpStream};
+use tracing::warn;
+
+use crate::{common::BuildError, ExporterFuture, PrometheusHandle};
+
+struct HttpListeningExporter {
+    inner: Arc<Inner>,
+}
+#[derive(Clone)]
+struct Inner {
+    handle: PrometheusHandle,
+    allowed_addresses: Option<Vec<IpNet>>,
+}
+
+impl HttpListeningExporter {
+    async fn serve(&self, listener: std::net::TcpListener) -> Result<(), hyper::Error> {
+        // TODO - address panic possibility
+        let listener = TcpListener::from_std(listener).unwrap();
+
+        loop {
+            let stream = match listener.accept().await {
+                Ok((stream, _)) => stream,
+                Err(e) => {
+                    warn!("Error accepting connection. Ignoring request. Error: {:?}", e);
+                    continue;
+                }
+            };
+
+            let remote_addr = match stream.peer_addr() {
+                Ok(remote_address) => remote_address.ip(),
+                Err(e) => {
+                    warn!("Error obtaining remote address. Ignoring request. Error: {:?}", e);
+                    continue;
+                }
+            };
+
+            self.process_stream(stream, remote_addr).await;
+        }
+    }
+
+    async fn process_stream(&self, stream: TcpStream, remote_address: IpAddr) {
+        let inner = self.inner.clone();
+        let service = service_fn(move |req: Request<body::Incoming>| {
+            let inner = inner.clone();
+            async move { Self::handle_http_request(&inner, remote_address, &req) }
+        });
+
+        tokio::task::spawn(async move {
+            if let Err(err) =
+                HyperHttpBuilder::new().serve_connection(TokioIo::new(stream), service).await
+            {
+                warn!("Error serving connection.  Error: {:?}", err);
```

Comment:
Add a dedicated `error` field on the event that we assign the error value to.

(see previous comments about logging for format example)